buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy

```diff
@@ -2587,40 +2587,49 @@ class BeamModulePlugin implements Plugin<Project> {
-        "group_id": project.name,
-        "java_expansion_service_jar": expansionJar,
-        "java_expansion_service_allowlist_file": javaClassLookupAllowlistFile,
-      ]
+      // initialize all expansion ports to -1. Will be populated in setupTask
+      def javaExpansionPorts = config.expansionProjectPaths.inject([:]) { map, k -> map[k] = -1; map }
+
       def usesDataflowRunner = config.pythonPipelineOptions.contains("--runner=TestDataflowRunner") || config.pythonPipelineOptions.contains("--runner=DataflowRunner")
       def javaContainerSuffix = getSupportedJavaVersion()
 
       // 1. Builds the chosen expansion service jar and launches it
       def setupTask = project.tasks.register(config.name+"Setup") {
         dependsOn ':sdks:java:container:' + javaContainerSuffix + ':docker'
-        dependsOn project.project(config.expansionProjectPath).shadowJar.getPath()
+        for (path in config.expansionProjectPaths) {
+          dependsOn project.project(path).shadowJar.getPath()
+        }
         dependsOn 'installGcpTest'
         if (usesDataflowRunner) {
           dependsOn ":sdks:python:test-suites:dataflow:py${project.ext.pythonVersion.replace('.', '')}:initializeForDataflowJob"
         }
         doLast {
-          project.exec {
-            // Prepare a port to use for the expansion service
-            javaExpansionPort = getRandomPort()
-            expansionServiceOpts.put("java_port", javaExpansionPort)
-            // setup test env
-            def serviceArgs = project.project(':sdks:python').mapToArgString(expansionServiceOpts)
-            executable 'sh'
-            args '-c', ". ${project.ext.envdir}/bin/activate && $pythonDir/scripts/run_expansion_services.sh stop --group_id ${project.name} && $pythonDir/scripts/run_expansion_services.sh start $serviceArgs"
+          // iterate through list of expansion service paths and build each jar
+          for (path in config.expansionProjectPaths) {
+            project.exec {
+              def expansionJar = project.project(path).shadowJar.archivePath
+              def javaClassLookupAllowlistFile = project.project(path).projectDir.getPath()
+              def expansionServiceOpts = [
+                "group_id": project.name,
+                "java_expansion_service_jar": expansionJar,
+                "java_expansion_service_allowlist_file": javaClassLookupAllowlistFile,
+              ]
+              // Prepare a port to use for the expansion service
+              javaExpansionPorts[path] = getRandomPort()
+              expansionServiceOpts.put("java_port", javaExpansionPorts[path])
+              // setup test env
+              def serviceArgs = project.project(':sdks:python').mapToArgString(expansionServiceOpts)
+              executable 'sh'
+              args '-c', ". ${project.ext.envdir}/bin/activate && $pythonDir/scripts/run_expansion_services.sh stop --group_id ${project.name} && $pythonDir/scripts/run_expansion_services.sh start $serviceArgs"
```

Comment:
Maybe we could defer this to a future CL, but possibly we could still use (and test) the auto-jar-expansion-service setup here rather than manually starting the services and injecting all the ports? @chamikaramj

Reply:
If we don't do this now, could you at least file a bug and drop a TODO.

Reply:
Ah sorry I thought this was a question for Cham. I'm unfamiliar with this "auto-jar-expansion-service" term, can you point me to somewhere I can learn about it?

Reply:
E.g. https://github.com/apache/beam/blob/v2.54.0-RC1/sdks/python/apache_beam/options/pipeline_options.py#L587

This is how, for example, https://github.com/apache/beam/blob/v2.54.0-RC1/sdks/python/apache_beam/transforms/sql.py#L86 works.

Reply:
Thanks for the suggestion, it cleaned up quite a bit of unnecessary code. 

It's now just building the relevant expansion jars and letting Python start them as needed. Let me know if this works! it's in the recent two commits